### PR TITLE
Authenticate

### PR DESCRIPTION
**Pull request by asledz**
## Description

Add Authenticate mechanism while creating new connection.
Add Authenticate request and possible responses to the frame.
Add serializers, deserializers.

## Fix

Fixes: #13 

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla/src/transport/connection.rs [hunk 1]
```diff
@@ -514,7 +519,36 @@ pub async fn open_named_connection(
     let result = connection.startup(options).await?;
     match result {
         Response::Ready => {}
-        Response::Authenticate => unimplemented!("Authentication is not yet implemented"),
+        Response::Authenticate(authenticate) => {
```
Comment (piodul) [inline scylla/src/transport/connection.rs:522]: We should return an error immediately if username and password is not configured. Now, we will send an invalid `AUTHENTICATE_RESPONSE` message (will contain a single `0` byte).

File: scylla/src/frame/response/authenticate.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+impl AuthSuccess {
+    pub fn deserialize(buf: &mut &[u8]) -> Result<Self, ParseError> {
+        let success_message = types::read_string(buf)?.to_string();
+
+        Ok(AuthSuccess { success_message })
+    }
+}
+
+#[derive(Debug)]
+pub struct AuthChallange {
```
[the remark this answers is not in the document]
Comment (asledz) [reply scylla/src/frame/response/authenticate.rs]: Done, I somehow always mix those two up :')

File: scylla/src/transport/connection.rs (hunk 1 above)
Comment (asledz) [reply scylla/src/transport/connection.rs:568]: Done! I was a bit confused because of cdrs implementation!

**Comment by asledz** [comment]
I somehow forgot to commit my changes on Friday 🤦‍♀️ But here is updated, working version, I just rebased it. So ready for review!

And I have a question about testing - I thought the best way is to add new scylla container with option `--authenticator AllowAllAuthenticator` and test on default superuser Cassandra (which I did) - I will add this container to the GitHub CI next thing tomorrow*, but now I am not so sure. Maybe we should create some sample users? Which option is better?

*In this commit I will also add a test that checks how authentication fails - so that all commits would pass checks separately 😅

**Comment by psarna** [comment]
Does this `cassandra` superuser have a password we can verify? If so, it's ok as an initial test, but it would be much better to prepare a scylla container with some pre-created users, and then test credentials for these users.

**Review by psarna — CHANGES_REQUESTED** [review]
Ok, after browsing the code I changed my mind above my statement above - let's get rid of the default `cassandra` user everywhere. Instead, the examples should work without authentication (and a separate auth example should be provided). Same for the test cases - the existing ones should either not be run if they detect that authentication is set up for the server *or* be able to handle such auth errors gently. A new separate set of tests should be prepared for authentication.

@piodul is there a neat way of marking some tests to be run only in specific conditions? E.g. we'd want authentication tests to be run only in workloads in which the server has proper authentication enabled.

**Comment by psarna** [comment]
Also, please make sure that if we get an `AUTHENTICATE` message from the server, but the authenticator is *not* password authenticator, then we end up with an error message that this authenticator type is not supported yet.

File: examples/cqlsh-rs.rs [hunk 3]
```diff
@@ -14,6 +14,7 @@ async fn main() -> Result<()> {
     let session: Session = SessionBuilder::new()
         .known_node(uri)
         .compression(Some(Compression::LZ4))
+        .user("cassandra", "cassandra")
```
[the remark this answers is not in the document]
Comment (asledz) [reply examples/cqlsh-rs.rs]: Fixed.

File: examples/parallel-prepared.rs [hunk 4]
```diff
@@ -11,7 +11,11 @@ async fn main() -> Result<()> {
 
     println!("Connecting to {} ...", uri);
 
-    let session: Session = SessionBuilder::new().known_node(uri).build().await?;
+    let session: Session = SessionBuilder::new()
+        .known_node(uri)
+        .user("cassandra", "cassandra")
+        .build()
+        .await?;
```
[the remark this answers is not in the document]
Comment (asledz) [reply examples/parallel-prepared.rs]: fixed

File: examples/parallel.rs [hunk 5]
```diff
@@ -11,7 +11,11 @@ async fn main() -> Result<()> {
 
     println!("Connecting to {} ...", uri);
 
-    let session: Session = SessionBuilder::new().known_node(uri).build().await?;
+    let session: Session = SessionBuilder::new()
+        .known_node(uri)
+        .user("cassandra", "cassandra")
+        .build()
+        .await?;
```
[the remark this answers is not in the document]
Comment (asledz) [reply examples/parallel.rs]: fixed

File: examples/select-paging.rs [hunk 6]
```diff
@@ -9,7 +9,11 @@ async fn main() -> Result<()> {
 
     println!("Connecting to {} ...", uri);
 
-    let session: Session = SessionBuilder::new().known_node(uri).build().await?;
+    let session: Session = SessionBuilder::new()
+        .known_node(uri)
+        .user("cassandra", "cassandra")
+        .build()
+        .await?;
```
[the remark this answers is not in the document]
Comment (asledz) [reply examples/select-paging.rs]: fixed

File: examples/user-defined-type.rs [hunk 7]
```diff
@@ -10,7 +10,11 @@ async fn main() -> Result<()> {
 
     println!("Connecting to {} ...", uri);
 
-    let session: Session = SessionBuilder::new().known_node(uri).build().await?;
+    let session: Session = SessionBuilder::new()
+        .known_node(uri)
+        .user("cassandra", "cassandra")
+        .build()
+        .await?;
```
[the remark this answers is not in the document]
Comment (asledz) [reply examples/user-defined-type.rs]: fixed

File: scylla/src/transport/connection_keeper.rs [hunk 8]
```diff
@@ -307,6 +307,8 @@ mod tests {
         let connection_config = ConnectionConfig {
             compression: None,
             tcp_nodelay: false,
+            auth_password: Some("cassandra".to_string()),
+            auth_username: Some("cassandra".to_string()),
```
[the remark this answers is not in the document]
Comment (asledz) [reply scylla/src/transport/connection_keeper.rs]: Ok!

File: examples/compare-tokens.rs [hunk 9]
```diff
@@ -10,7 +10,11 @@ async fn main() -> Result<()> {
 
     println!("Connecting to {} ...", uri);
 
-    let session: Session = SessionBuilder::new().known_node(uri).build().await?;
+    let session: Session = SessionBuilder::new()
+        .known_node(uri)
+        .user("cassandra", "cassandra")
```
[the remark this answers is not in the document]
Comment (asledz) [reply examples/compare-tokens.rs]: fixed

**Comment by asledz** [comment]
I found a way to run some tests separately in GitHub in a different container - for example with some users specified. I wanted to consult scheme for test I came up before pushing them to this PR:
- for each authentication type there would be a separate test
- each test would consider of following:
-- create a connection with superuser Cassandra and create sample user
-- create new connection with sample user
-- failed attempt to connect with invalid credentials
- each test in GitHub would run as a different check (since we need a different container for them)
Does it sound like a good idea, or there are more test cases we want to cover?

Also - I just added all of Scylla supported authentication types, will push implementation for missing ones in another commit tomorrow 😅

**Comment by psarna** [comment]
There seems to be a conflict with current `main` - please rebase on top of it (preferably with `git pull --rebase` to avoid merges).

**Comment by psarna** [comment]
The code looks good. After fixing the CI issue:
```

.github/workflows/authenticate_test.yml#L24
The workflow is not valid. .github/workflows/authenticate_test.yml (Line: 24, Col: 9): Unexpected value 'authenticator' .github/workflows/authenticate_test.yml (Line: 30, Col: 3): 'build' is already defined

```
it looks good to go.

**Comment by psarna** [comment]
@cvybhu could you assist with getting CI to work with authenticators?

**Comment by psarna** [comment]
ping

**Comment by cvybhu** [comment]
@psarna

We didn't find any way to enable authentication for scylla service nicely in the CI
* There doesn't exist any way to pass arguments to docker run
* Mounting files doesn't really work because services start before the checkout action and the files are not there yet
* We could set some environment variable but we couldn't find any that would enable authentication

We will just have to create a custom image on Dockerhub :/

**Comment by psarna** [comment]
That's alright, I was actually sure that a custom docker image is the way we agreed to do it in the first place :smile:

**Comment by cvybhu** [comment]
@psarna CI is working now

**Pull request merged**